### ISSUE TITLE
Simplify/Clarify Web site building

### ISSUE THREAD

**Issue by Drup** [ocaml/dune#663]
I think this use case deserves to be better explained, maybe in the "quickstart" section. Currently there are many things that are not very clear:
- An executable with `(js_of_ocaml (..))` and a public name will build/install the `.exe`, not the `.js`. This is not intuitive.
- If I have an HTML file that uses the generated js, it's not clear how to put the website together in order to visit it without either pointing inside `_build` or moving the files around.
- What's the right way to handle static data and/or bundle additional js files?

I feel like the basic supports is there (and I ended up successfully building my `.bc.js`) but it feel fairly unfinished. I'm not sure what the idiomatic dune way is, but a little bit of polish would be sufficient to make the situation much better. Maybe with a new target that would put the files together and launch the website, or something like that ?

**Comment by diml** [comment]
I'm for improving support for building web sites, but I don't know much about web development. Could you write a small description of what better would look like?

/cc @hhugo who wrote the jsoo rules

**Comment by Drup** [comment]
So, for a very basic static web site, the situation will at least look like that:
- Some executable that get compiled to a js file.
- Some HTML/CSS/Other files that reference the generated js file and each others. These files can be either static (and then sit in `src/` or `data/`), or compilation products.

That's already annoying to deal with, since the compilation products are lost somewhere in `_build`, so the relative path will not agree. You also want all the files to be grouped to move them for deployment.

That's for a static website. A dynamic website can be more complicated, but the path issues will be similar.

**Comment by rgrinberg** [comment]
> An executable with (js_of_ocaml (..)) and a public name will build/install the .exe, not the .js. This is not intuitive.

Yes, this is because `js_of_ocaml` is only for passing in specific jsoo options. The jsoo rules are always setup. What would make sense with the way things work is perhaps to add `jsoo` to `(modes (...))`. Since that's the option where we control what `.exe` actually becomes.

> If I have an HTML file that uses the generated js, it's not clear how to put the website together in order to visit it without either pointing inside _build or moving the files around.

How could you avoid referring to `_build`? With few exceptions, dune sticks to writing build artifacts into that directory. What could possibly make sense it copy your html to `_build/<context>` as well so that you can don't have to include `_build` in your relative paths.

> What's the right way to handle static data and/or bundle additional js files?

`(js_of_ocaml (javascript_files ..))` exists but I'm not sure that's what you're looking for.

I'm actually not sure that adding web dev support to dune makes a lot of sense. Unless it's something really simple. There are better tools out there like webpack that will handle preparing static data for a website much better.

@jordwalke do you have any ideas how dune could work better in web contexts? In particular on the client side with jsoo.

**Comment by jordwalke** [comment]
I tried out the Dune/jsoo integration for my [`flex`](https://github.com/jordwalke/flex) benchmarking experiment, and I could tell that it has come a long way but I agree with @Drup that the docs haven't caught up.

My general feedback is that the `.bc.js` is actually okay. The fact that the artifacts all end up in `_build` is actually okay. Believe it or not the fact that jbuilder+jsoo puts all the artifacts of its dependencies into the top level project's root (in `_build`) might sound like a drawback, but it's actually required for correctness in an npm environment where symlinks abound. You can't put your dependencies' artifacts into *their* package/install locations.

The only things that I observed needed improvement were the following:
- Docs for what works right now.
- Have the JS output adhere to commonJS format (at least in the separate compilation mode). That means you lay out modules into a hierarchy on disk and from one module they import another module by doing `var OtherModule = require('./relative/path/to/other/module')`.
- Some improvements in sourcemaps correctness.

Even if you were to document how to setup a basic project with some dumb html files that require the JS files - outputting modules that just mutate a global_data object runs so contrary to common web workflows that it will loose many web developers. Still, documenting the way things work right now would be great for non-web people. Once you output to a commonJS format, there's nothing to document for web developers besides the jbuilder config/usage. Everything will just work with their current ways of running JS in web pages (that means [webpack](https://github.com/webpack/webpack)). That's the benefit of targeting the most popular standard.

It shouldn't be hard to implement commonJS output instead of mutating the global_data object (in separate compilation mode). Just try to make each OCaml top level module appear in its own file, in its own directory like:

```
_build/library_foo/module_abc/index.js
_build/library_bar/module_xyz/index.js
```

Then instead of just mutating / reading global data, `module_xyz` from `library_bar`, can pull in another library/module by doing `require('library_foo/module_abc')`. `./relative/path/index.js` is the module that is pulled in by default when doing `require('./relative/path/')`.

Then people just point their webpack config to the `_build` directory and everything just works like they are used to!

**Comment by diml** [comment]
Thanks for the feedback @jordwalke. /cc @hhugo about the commonJS format, I believe this would require some changes in jsoo

**Comment by Drup** [comment]
I think something much simpler would be enough for most use cases: You specify that a bunch of static files should be placed alongside the build artifacts, maybe specify some structure. Jbuilder would tell you where to find the entry point, and that's it. That could probably be useful for other things as well (unikernels, for example)

**Comment by diml** [comment]
I suppose we already have a similar mechanism for installed files: the user marks some files with a package name + section and they end up at the right place in `_build/install`.

We could generalize this idea. For instance you'd mark the files you want to include in your website and jbuilder would create `_build/www` with the right structure. I don't know how this should be called, maybe *bundle*:

```scheme
(bundle
 ((name www)
  (files (index.html (main.bc.js as main.js)))))
```

**Comment by Drup** [comment]
@diml Yes, this is exactly the kind of things that I had in mind. This could produce a new target as well, so programmers would do `jbuilder build @www` and have everything ready.

**Comment by Alizter** [comment]
@anmonteiro Any opinion on whether this would be a useful feature?
